### Import required dependencies

In [86]:
import pandas as pd
import numpy as np
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [87]:
# Create the path and import the data
student_data = os.path.join('./Resources', 'new_full_student_data.csv')
student_df = pd.read_csv(student_data)

In [88]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [89]:
# Check for null values by using isna() or isnull()
#cln_df = student_df.isnull().sum()
cln_df = student_df.isna().sum()
cln_df

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [90]:
# Number of rows with at least one null
nan_rows = len(student_df.loc[student_df.isnull().any(axis=1)].index)
display("NaN rows=", nan_rows)

'NaN rows='

2847

In [91]:
# Drop rows with null values and verify removal
cln_df = student_df.dropna(how='any')
cln_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [92]:
# Check for duplicated rows
duplicate_rows = cln_df.duplicated().sum()
clean_rows = cln_df.count() - duplicate_rows
display("Duplicate rows=", duplicate_rows, "Clean rows=", clean_rows['reading_score':'math_score'])

'Duplicate rows='

1836

'Clean rows='

reading_score    14831
math_score       14831
dtype: int64

In [93]:
# Drop duplicated rows and verify removal
cln_df = cln_df.drop_duplicates()
duplicate_rows = cln_df.duplicated().sum()
display("Duplicate rows=", duplicate_rows)

'Duplicate rows='

0

In [94]:
# Check data types
cln_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [95]:
# Examine the grade column to understand why it is not an int
cln_df['grade']

0         9th
1         9th
2         9th
3         9th
5         9th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 14831, dtype: object

In [96]:
# Remove the non-numeric characters and verify the contents of the column
cln_df.loc[:, 'grade'] = cln_df['grade'].str.replace('th', '')
cln_df.loc[:, 'grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [97]:
# Change the grade column to the int type and verify column types,
# use np.int64 instead of int to align with image in the instruction
#cln_df['grade'] = cln_df['grade'].astype(int)
cln_df['grade'] = cln_df['grade'].astype(np.int64)
cln_df.dtypes

student_id         int64
student_name      object
grade              int64
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [98]:
# Display summary statistics for the DataFrame
cln_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [99]:
# Display the mean math score using the mean function
ave_math_score = cln_df.loc[:, 'math_score'].mean()
display("Average math score=", ave_math_score)

'Average math score='

64.67573326141189

In [100]:
# Store the minimum reading score as min_reading_score
max_reading_score = cln_df['reading_score'].max()
min_reading_score = cln_df['reading_score'].min()
display("Max reading score=", max_reading_score, "Min reading_score=", min_reading_score)

'Max reading score='

100.0

'Min reading_score='

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [101]:
# Use loc to display the grade column
cln_df.loc[:, 'grade']

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: int64

In [102]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
cln_df.iloc[:3, 3:6]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [103]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
cln_df.loc[cln_df['grade'] == 9].describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [104]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
#max_reading_row = cln_df.loc[cln_df['reading_score'] == max_reading_score]
min_reading_row = cln_df.loc[cln_df['reading_score'] == min_reading_score]
#display(max_reading_row, min_reading_row)
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [105]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
dixon_hs_gr10_df = cln_df.loc[(cln_df['school_name'] == "Dixon High School") & (cln_df['grade'] == 10), ['school_name', 'reading_score',]]
dixon_hs_gr10_df

school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[569 rows x 2 columns]

In [106]:
# Find the mean reading score for all students in grades 11 and 12 combined.
ave_reading_score_gr1112 = cln_df.loc[(cln_df['grade'] >= 11), ['reading_score']].mean()
ave_reading_score_gr1112

reading_score    74.900381
dtype: float64

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

### Deliverable 5 Requirements (cited from instructions on Module 4 Challenge)
- The average budget for each school type is displayed. (7 points)

- The total number of students per school is displayed in descending order. (6 points)

- The average math scores for each combination of grade and school type are displayed. (7 points)

In [107]:
# Double check using loc conditionals and mean to find the average reading and math scores for each school type.
charter_hs_df = cln_df.loc[cln_df['school_type'] == "Charter", ['reading_score', 'math_score', 'school_budget']].mean()
pub_hs_df = cln_df.loc[cln_df['school_type'] == "Public", ['reading_score', 'math_score', 'school_budget']].mean()
display("Charter:", charter_hs_df, "Public:", pub_hs_df)

'Charter:'

reading_score        72.450603
math_score           66.761883
school_budget    872625.656236
dtype: float64

'Public:'

reading_score        72.281219
math_score           62.951576
school_budget    911195.558251
dtype: float64

In [108]:
# Use groupby and mean to find the average reading and math scores for each school type.
ave_student_scores_by_grade = cln_df.groupby('grade').mean()
ave_student_scores_by_grade['school_budget'] = ave_student_scores_by_grade['school_budget'].map("${:,.2f}".format)
display(ave_student_scores_by_grade.loc[:, ['reading_score', 'math_score', 'school_budget']])
ave_student_scores_by_district = cln_df.groupby('school_type').mean()
ave_student_scores_by_district['school_budget'] = ave_student_scores_by_district['school_budget'].map("${:,.2f}".format)
display(ave_student_scores_by_district.loc[:, ['reading_score', 'math_score', 'school_budget']])

reading_score  math_score school_budget
grade                                         
9          69.236713   66.585624   $898,692.61
10         71.647590   64.911778   $896,341.54
11         77.478417   64.204911   $885,658.60
12         72.245103   62.283795   $891,797.76

reading_score  math_score school_budget
school_type                                         
Charter          72.450603   66.761883   $872,625.66
Public           72.281219   62.951576   $911,195.56

In [109]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
count_by_school = cln_df.groupby('school_name').count()
count_by_school['student_count'] = count_by_school['student_id']
count_by_school.loc[:, ['student_count']].sort_values(by='student_count', ascending = False)

student_count
school_name                          
Montgomery High School           2038
Green High School                1961
Dixon High School                1583
Wagner High School               1541
Silva High School                1109
Woods High School                1052
Sullivan High School              971
Turner High School                846
Bowers High School                803
Fisher High School                798
Richard High School               551
Campos High School                541
Odonnell High School              459
Campbell High School              407
Chang High School                 171

In [110]:
ave_by_schooltype_grade = cln_df.groupby(['school_type', 'grade']).mean()
ave_by_schooltype_grade['math_score'] = ave_by_schooltype_grade['math_score'].map("{:,.1f}".format)
ave_by_schooltype_grade['school_budget'] = ave_by_schooltype_grade['school_budget'].map("${:,.2f}".format)
display(ave_by_schooltype_grade.loc[:, ['math_score', 'school_budget']])
# Use np.round to match image in the instruction
ave_by_schooltype_grade = cln_df.groupby(['school_type', 'grade']).mean().apply(np.round)
display(ave_by_schooltype_grade.loc[:, ['math_score']])

math_score school_budget
school_type grade                         
Charter     9           70.1   $863,817.29
            10          66.4   $871,823.61
            11          68.0   $874,262.71
            12          60.2   $885,096.34
Public      9           63.8   $926,800.16
            10          63.8   $914,715.36
            11          59.3   $900,248.91
            12          63.6   $895,952.92

math_score
school_type grade            
Charter     9            70.0
            10           66.0
            11           68.0
            12           60.0
Public      9            64.0
            10           64.0
            11           59.0
            12           64.0

In [111]:
# Export the clean data without row index
student_data = os.path.join('./Resources', 'new_full_student_data_clean.csv')
cln_df.to_csv(student_data, index=False)

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

## Results and Summary
Let us assess our results and summary that meet the requirements of each deliverable in the following sections.

### Deliverable 1 and 2 Summary
All data collection, preparation and verification steps are properly done by using `os.path.join()`, `pd.read_csv()`, etc. In Module 4 assignment, we imported student data as a pandas DataFrame. A few examples that might be unique when preparing and cleaning the data are summarized below.

- After data cleaning steps, I copied the original `student_df` DataFrame as `cln_df`, a step that may be different from others'. I then crunched the clean version of the DataFrame based on `cln_df`. This allowed me to save the clean datasets and validate the accuracy of our data manipulation steps when necessary.
- Added a few checkpoints for making sure every step was accurately performed. I mainly used `display()` for printing the analysis results.
  - 2847 lines containing at least one null were removed.
  - 1836 duplicated rows were trimmed.

```
# Drop duplicated rows and verify removal
cln_df = cln_df.drop_duplicates()
duplicate_rows = cln_df.duplicated().sum()
display("Duplicate rows=", duplicate_rows)
```

```
# use np.int64 instead of int to align with image in the instruction
cln_df['grade'] = cln_df['grade'].astype(np.int64)
```

### Deliverable 3 Summary
The average math score for all students was **64.67573326141189**, which matched with the statistics displayed by `describe()`. We calculated the max reading and min reading scores by applying `max()` and `min()` functions, and stored the max_reading_score as well as min_reading_score (**10.5**) values for later use.

### Deliverable 4 Summary
We applied either `loc` or `iloc` functions to analyze specific subsets based on certain condition(s). We also used Boolean operators, like `&`, `|`, or `~`, to filter, index or slice, and select or group specific subsets. We were able to analyze all the required datasets and our results matched well with the reference data.

The mean reading score for all students in grades 11 and 12 combined was **74.900381**. I adopted a simplified conditional statement after verifying that there were no grades beyond grade 12<sup>th</sup> in our datasets.

```
# Find the mean reading score for all students in grades 11 and 12 combined.
ave_reading_score_gr1112 = cln_df.loc[(cln_df['grade'] >= 11), ['reading_score']].mean()
ave_reading_score_gr1112
```

### Deliverable 5 Summary
Since we noticed two conflicting versions of deliverables, I printed reading and math scores in addition to the requested school budget by school type altogether. To assure our results, I also used `loc` along with conditional statements to reconfirm that our data and calculation were solid. Now I feel comfortable providing whatever data immediately that the chief data scientist for a city school district may further inquire.

- Average test scores and school funding per school type are shown in **Table 1**. Overall test scores were slightly higher for Charter schools in general even though they received lower funding.
- Average test scores and school funding by grade level are displayed in **Table 2** for comparison purposes. Math scores were in decreasing trends as grades moved from lower to higher.
- To align with the best practice approach, school budget was rounded to the nearest whole cent in both tables by applying extra `map()` formatting.

<p style="text-align: center;"><b>Table 1. Average Test Scores and Funding by School Type</b></p>

| school_type    | reading_score | math_score | school_budget |
| :---           |  ---:         |  ---:      |  ---:         |
| <b>Charter</b> | 72.450603     | 66.761883  | \$872,625.66  |
| <b>Public</b>  | 72.281219     | 62.951576  | \$911,195.56  |

<p style="text-align: center;"><b>Table 2. Average Test Scores and Funding by Grade</b></p>

| grade | reading_score | math_score | school_budget |
| :---  |  ---:         |  ---:      |  ---:         |
|  9    | 69.236713     | 66.585624  | \$898,692.61  |
| 10    | 71.647590     | 64.911778  | \$896,341.54  |
| 11    | 77.478417     | 64.204911  | \$885,658.60  |
| 12    | 72.245103     | 62.283795  | \$891,797.76  |

The total number of students per school in descending order was calculated by using `groupby`, `count`, and `sort_values` as shown in the following short code. To retain the clean DataFrame, I added a column named *student_count*, which contained an exact copy of student counts from column *student_id* or any existing column besides *school_name* (**Table 3**).

```
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.
count_by_school = cln_df.groupby('school_name').count()
count_by_school['student_count'] = count_by_school['student_id']
count_by_school.loc[:, ['student_count']].sort_values(by='student_count', ascending = False)
```

<p style="text-align: center;"><b>Table 3. Total Number of Students per School</b></p>

| school_name            | student_count |
| :---                   |          ---: |
| Montgomery High School | 2038 |
| Green High School      | 1961 |
| Dixon High School      | 1583 |
| Wagner High School     | 1541 |
| Silva High School      | 1109 |
| Woods High School      | 1052 |
| Sullivan High School   |  971 |
| Turner High School     |  846 |
| Bowers High School     |  803 |
| Fisher High School     |  798 |
| Richard High School    |  551 |
| Campos High School     |  541 |
| Odonnell High School   |  459 |
| Campbell High School   |  407 |
| Chang High School      |  171 |

The average math scores (and school budget) for each combination of grade and school type, which were analyzed by using `groupby` and `mean`, are shown in **Table 4** and **Table 5**. **Table 5** showed the results when we added `apply(np.round)`, such that numbers matched well with image in the instruction of Module 4 assignment.

```
# Use np.round to match image in the instruction
ave_by_schooltype_grade = cln_df.groupby(['school_type', 'grade']).mean().apply(np.round)
display(ave_by_schooltype_grade.loc[:, ['math_score']])
```

<p style="text-align: center;"><b>Table 4. Average Math Scores and School Budget per Grade for each School Type</b></p>

| school_type    | grade | math_score | school_budget  |
| :---           | ---:  | ---:       | ---:           |
| <b>Charter</b> | 9     | 70.1       | \$863,817.29   |
|                | 10    | 66.4       | \$871,823.61   |
|                | 11    | 68.0       | \$874,262.71   |
|                | 12    | 60.2       | \$885,096.34   |
| <b>Public</b>  | 9     | 63.8       | \$926,800.16   |
|                | 10    | 63.8       | \$914,715.36   |
|                | 11    | 59.3       | \$900,248.91   |
|                | 12    | 63.6       | \$895,952.92   |

<p style="text-align: center;"><b>Table 5. Average Math Scores per Grade for each School Type</b></p>

| school_type    | grade | math_score |
| :---           | ---:  | ---:       |
| <b>Charter</b> | 9     | 70.0       |
|                | 10    | 66.0       |
|                | 11    | 68.0       |
|                | 12    | 60.0       |
| <b>Public</b>  | 9     | 64.0       |
|                | 10    | 64.0       |
|                | 11    | 59.0       |
|                | 12    | 64.0       |

As a final checkpoint, I exported the clean data into a csv file and rerun the refactored Jupyter Notebook code for making sure everything has been analyzed correctly.

```
# Export the clean data without row index
student_data = os.path.join('./Resources', 'new_full_student_data_clean.csv')
cln_df.to_csv(student_data, index=False)
```
